facebook/Docusaurus
# feat(v2): proper realtime hot reload with react-hot-loader

## Motivation

## TL;DR See Test Plan

@yangshun mentioned to me on Discord that the hot-reload is actually not working. So I did some investigation and it was indeed not working. It fallbacks to page refresh.

It was because I forgot to correctly configure the `module.hot` 😢.  See https://webpack.js.org/guides/hot-module-replacement/.

Anyway. I learned a lot about Hot Module Replacement. Things are slightly more complicated on our side because ***in development: we are a SPA*** and ***in production: we are SSR + SPA*.** There is a reason why tools like `create-react-app` is very popular, hahaha. They settle this kind of things for us

However, I remembered [Dan Abramov - Live React: Hot Reloading with Time Travel at react-europe 2015](https://www.youtube.com/watch?v=xsSnOQynTHs), so in this PR I added https://github.com/gaearon/react-hot-loader to take the hot reload into next level. (Real-time hot reload)



Other changes:
- Add `--hot-only` CLI options to let user disable fallback to full page refresh if hot-reload is not working
- Use express.static to follow v1 way, contentBase should be the default publicPath
- Add some TODO plugin hooks on devserver

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

### Development:

**Before**

Notice that there is a flash of screen. This is in fact a full page refresh. This is quite similar to v1 way of live reload. 
![not hot reload](https://user-images.githubusercontent.com/17883920/55548204-8b98a300-5705-11e9-8db4-9413260ad6ee.gif)

**After**

Notice that there is no flash of screen. This is a hot reload. In fact, it's realtime
![hot reload fix](https://user-images.githubusercontent.com/17883920/55548221-96533800-5705-11e9-8325-2365130eed94.gif)

### Production:
Netlify - https://deploy-preview-1336--docusaurus-2.netlify.com/
 
## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)
